adding --default-proxy flag to use the system proxy

evolu1 opened the pull request:
In order to make API requests successful in working environments with proxy configurations, it is necessary to use the system proxy in HttpClient. 

The flag `--default-proxy` is added. When used, the system proxy is applied for every http request.

Implemented using https://docs.microsoft.com/de-de/dotnet/api/system.net.http.httpclient.defaultproxy

tomchavakis commented:
Hi @evolu1 ,

Your contribution is really important. 
Can you add a test on methods and/or an example to the documentation?
I am not aware of the usage of the Proxy credentials and CredentialCache.DefaultCredentials.
Can you explain how this works shorty, please?
Thanks in advance

evolu1 commented:
I put some work into this and I have to correct myself a bit 🙃

Included in the last commit:

- added `--proxy-url` to support proxy server usage
- replaced `--default-proxy` with `--proxy-system-auth` as it is more fitting
- added example usage in readme

---

> Can you add a test on methods and/or an example to the documentation?

I can add an example yes. It is just a bool. Setting the flag enables it.
**I am not familiar with c# unit testing at all, to be honest.**

I also took a look at the older PR #77 and added the functionality behind it, setting a proxy url.

> Can you explain how this works shorty, please?

It is common in working environments to use a proxy server to pass all outgoing traffic through it. Usually because of security reasons, to whitelist traffic and so on.

A proxy is configured on a machine with a protocol (http), an ip/hostname and a port. 
Example: `http://my.proxy.com:8080`
So PR #77 would work like `--proxy "http://my.proxy.com:8080"`
This is now called `--proxy-url "..."`
Sometimes there is authentication too, which is not accounted for in the above. 
Authentication for the proxy server could also be added to nuget-license but that would mean that username and password needs to be passed in cleartext as a flag argument. This is not safe at all and I do not know how to that securely. 
However authentication credentials can be safely read in from the system...

---

My PR here #95 focuses on this problem:

A machine in a working environment needs to make the http requests of nuget-license but can't, because the HttpClient does not use the correct proxy server (outgoing connection blocked by proxy). 
With this PR a proxy server URL can be set (`--proxy-url`), so that a proxy server will be used for all http requests. Just like in #77.
This is working if the proxy server does not require authentication.

If authentication is required, then `--proxy-system-auth` will read in the system credentials from the current context in which the application (nuget-license) is running and use these credentials to authenticate at the proxy server ([CredentialCache.DefaultCredentials Property](https://docs.microsoft.com/en-us/dotnet/api/system.net.credentialcache.defaultcredentials)).
Meaning, under Windows, the Windows user and password will be used to authenticate at the proxy server. This is solution I need, safely reading in credentials and using them for the proxy server.

tomchavakis commented:
Hi @evolu1,

I would like to thank you for your detailed response. Now it's clear and I think that the implementation is better than previously.
Regarding unit tests, I will add them previously and let you know when it's done and will add you as a reviewer in order to take a look at the implementation.
I will create a new tag in order to be able to check it up. If it's possible please make sure that is running and there is no issue with your proxy environment.

Thank's again for your pull request.
